### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by shreyash-x**
Code Review

### CODE REVIEW

File: experiment-descriptor.json [hunk 1]
```diff
@@ -1,53 +1,67 @@
...
+    },
+    {
+      "unit-type": "task",
+      "label": "Pretest",
+      "content-type": "assesment",
+      "source": "pretest.json",
+      "target": "pretest.html"
+    },
+    {
+      "target": "demo-RSFF.html",
```
Comment (raj-vlabs) [inline experiment-descriptor.json]: can we just call it demo.html since there's no other demo file anyway

File: experiment/objective.md [hunk 2]
```diff
@@ -1,3 +1,3 @@
-1. The objective of this experiment is to understand the working of the Flip Flops.
-2. Try to make basic Flip Flops like JK, RS , T , D Flip Flops.
-
+1. The objective of this experiment is to understand the working of the Flip Flops.
+2. Try to make basic Flip Flops like JK, RS , T , D Flip Flops.
```
Comment (raj-vlabs) [inline experiment/objective.md]: The second point can be removed, so can the numbering of the first point.

File: experiment/posttest.json [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                "b": "SET",
+                "c": "RESET",
+                "d": "TOGGLE"
+            },
+            "correctAnswer": "a",
+            "explanations": {},
+            "difficulty": "beginner"
+        },
+        {
+            "question": "2. A NAND gate is 0 if and only if ?",
```
Comment (raj-vlabs) [inline experiment/posttest.json]: ? is not needed at the end of the question

File: experiment/posttest.json [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            "answers": {
+                "a": "Both inputs are 0",
+                "b": "Atleast one is 0",
```
Comment (raj-vlabs) [inline experiment/posttest.json]: Space between 'At least'

File: experiment/posttest.json [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                "d": "SET=1 and RESET=1"
+            },
+            "correctAnswer": "d",
+            "explanations": {
+                
+            },
+            "difficulty": "intermediate"
+        },
+        {
+            "question": "4. A NAND gate can be formed by ?",
```
Comment (raj-vlabs) [inline experiment/posttest.json]: ? is not needed at the end of the question

File: experiment/pretest.json [hunk 6]
```diff
@@ -0,0 +1,75 @@
+{
+    "version": 2.0,
+    "questions": [
+        {
+            "question": "1. The output of a sequenrtial circuit depends on ?",
```
Comment (raj-vlabs) [inline experiment/pretest.json]: 1. Sequential is mis-spelt
2. ? is not needed at the end of the question

File: experiment/pretest.json [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                "d": "None of these"
+            },
+            "correctAnswer": "c",
+            "explanations": {
+                
+            },
+            "difficulty": "beginner"
+        },
+        {
+            "question": "2. What id the full form of SR flip flop ?",
```
Comment (raj-vlabs) [inline experiment/pretest.json]: What is the full form
in place of
What id the full form

File: experiment/pretest.json [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                "d": "4"
+            },
+            "correctAnswer": "d",
+            "explanations": {
+                
+            },
+            "difficulty": "Intermediate"
+        },
+        {
+            "question": "The JK flip flop have a _________ memory ? ",
```
Comment (raj-vlabs) [inline experiment/pretest.json]: The JK flip flop has a

File: experiment/pretest.json [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+                "d": "True"
+            },
+            "correctAnswer": "d",
+            "explanations": {
+               
+            },
+            "difficulty": "Intermediate"
+        },
+        {
+            "question": "5. THe flip-flop is a/an _________ device ?",
```
Comment (raj-vlabs) [inline experiment/pretest.json]: The
in place of 
THe

File: experiment/simulation/css/main.css [hunk 10]
```diff
@@ -1 +1,459 @@
...
+
+.nand {
+  background-image: url("../images/nand.svg");
+}
+
+.nor {
+  background-image: url("../images/nor.svg");
+}
+
+.threeipnand {
```
Comment (raj-vlabs) [inline experiment/simulation/css/main.css]: three-ip-nand

File: experiment/simulation/demo-DFF.html [hunk 11]
```diff
@@ -0,0 +1,134 @@
+<!DOCTYPE html>
+<html>
```
Comment (raj-vlabs) [inline experiment/simulation/demo-DFF.html]: Duplicate html tag

File: experiment/simulation/js/animation-DFF.js [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+
+function appendData() {
+    if (textInput[0].textContent !== "0" && timeline.progress() === 0) {
+        changeto0(16,124,0,0);
+        observ.innerHTML = "Data bit is equal to 0";
+    }
+    else if (textInput[0].textContent !== "1" && timeline.progress() === 0) {
+        changeto1(16,124,0,0);
+        observ.innerHTML = "Data bit is equal to 1";
+    }
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-DFF.js:148]: Please see if a default else block is needed

File: experiment/simulation/js/animation-DFF.js [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+    setter(textInput[0].textContent,dots[1]);
+}
+function appendClock() {
+    if (textInput[1].textContent !== "0" && timeline.progress() === 0) {
+        changeto0(16,514,1,1);
+    }
+    else if (textInput[1].textContent !== "1" && timeline.progress() === 0) {
+        changeto1(16,514,1,1);
+        observ.innerHTML = "Clock is turned ON";
+    }
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-DFF.js:163]: Please see if a default else block is needed

File: experiment/simulation/js/animation-DFF.js [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+
+
+    for(const text of textInput){
+        if (text.textContent === "2") {
+            observ.innerHTML = "Please set the input values";
+            return;
+        }
+    }
+    if (timeline.progress() !== 1) {
+        if (!circuitStarted) {
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-DFF.js]: Do we need the condition check here?

File: experiment/simulation/js/flipflop.js [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+import { setPosition } from "./layout.js";
+import { gates,printErrors } from './gate.js';
+
+'use strict';
+
+export let flipFlops = {};
+
+export function clearFlipFlops() {
+
+    for (let ffID in flipFlops) {
```
Comment (raj-vlabs) [inline experiment/simulation/js/flipflop.js]: Do we need this block? Generally it's not a good idea to delete from a collection while iterating over it

File: experiment/simulation/js/gate.js [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+    // logic gates and flip flop
+    for (let iterations = 0; iterations < 5; iterations++) {
+        for (let gateId in gates) {
+            const gate = gates[gateId];
+            if (gate.isOutput === false && gate.isInput === false && gate.type != "NOT" && gate.type != "ThreeIPNAND") {
+                const val1 = getOutput(gate.inputs[0]);
+                const val2 = getOutput(gate.inputs[1]);
+                if (val1 == null || val2 == null) {
+                    let val = null;
+                    if (val1 == null && val2 == null) {
```
Comment (raj-vlabs) [inline experiment/simulation/js/gate.js]: Please use ===

File: experiment/simulation/js/layout.js [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+  }
+
+  if (window.currentTab != null) {
+    document.getElementById(window.currentTab).classList.remove("is-active");
+  }
+  window.currentTab = task;
+  document.getElementById(task).classList.add("is-active");
+
+  // Half adder
+  if (task === "task1") {
```
Comment (raj-vlabs) [inline experiment/simulation/js/layout.js]: Please use switch-case for better readability

File: experiment/simulation/js/layout.js [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+                <th>Q</th>
+                <th>Q'</th>
+                <th>Q</th>
+                <th>Q'</th>
+              </tr>
+            </thead>`
+  }
+  else
+  {
+    console.log("Error: Unknown tab");
```
Comment (raj-vlabs) [inline experiment/simulation/js/layout.js]: Please use console.debug where required

**Review by raj-vlabs — CHANGES_REQUESTED** [review]

File: experiment-descriptor.json (hunk 1 above)
Comment (shreyash-x) [reply experiment-descriptor.json]: Yes we can, had it named like that to identify which circuit it has, changed it now.

File: experiment/objective.md (hunk 2 above)
Comment (shreyash-x) [reply experiment/objective.md]: Done

File: experiment/posttest.json (hunk 3 above)
Comment (shreyash-x) [reply experiment/posttest.json]: Fixed

File: experiment/posttest.json (hunk 4 above)
Comment (shreyash-x) [reply experiment/posttest.json]: Done

File: experiment/posttest.json (hunk 5 above)
Comment (shreyash-x) [reply experiment/posttest.json]: Done wherever required

File: experiment/pretest.json (hunk 6 above)
Comment (shreyash-x) [reply experiment/pretest.json]: Fixed it

File: experiment/pretest.json (hunk 7 above)
Comment (shreyash-x) [reply experiment/pretest.json]: Done

File: experiment/pretest.json (hunk 8 above)
Comment (shreyash-x) [reply experiment/pretest.json]: Done

File: experiment/pretest.json (hunk 9 above)
Comment (shreyash-x) [reply experiment/pretest.json]: Done

File: experiment/simulation/css/main.css (hunk 10 above)
Comment (shreyash-x) [reply experiment/simulation/css/main.css]: fixed

File: experiment/simulation/demo-DFF.html (hunk 11 above)
Comment (shreyash-x) [reply experiment/simulation/demo-DFF.html]: Removed

File: experiment/simulation/js/animation-DFF.js (hunk 13 above)
Comment (shreyash-x) [reply experiment/simulation/js/animation-DFF.js:163]: Will have to leave the else block empty, it is not an error, we just can't change the values when timeline.progress isn't 0.

Comment (shreyash-x) [reply experiment/simulation/js/animation-DFF.js:163]: Added the else block, it displays a message in the observation table

File: experiment/simulation/js/animation-DFF.js (hunk 14 above)
Comment (shreyash-x) [reply experiment/simulation/js/animation-DFF.js]: No need, removed it

File: experiment/simulation/js/flipflop.js (hunk 15 above)
Comment (shreyash-x) [reply experiment/simulation/js/flipflop.js]: Just assigned the object to {}

File: experiment/simulation/js/gate.js (hunk 16 above)
Comment (shreyash-x) [reply experiment/simulation/js/gate.js]: Done

File: experiment/simulation/js/layout.js (hunk 17 above)
Comment (shreyash-x) [reply experiment/simulation/js/layout.js]: Done

File: experiment/simulation/js/layout.js (hunk 18 above)
Comment (shreyash-x) [reply experiment/simulation/js/layout.js]: Done

**Review by raj-vlabs — APPROVED** [review]

**Pull request merged**